### PR TITLE
Tolerate packet reordering in the early init process

### PR DESCRIPTION
**Pull request by est31**
Fixes a bug where packet reordering made the server give the
client two peer ids instead of one. This in turn confused
reliable packet sending and made connecting to the server fail.

For my detailed theory how this happens and further discussion see the commit msg.

However, marked WIP, as I still don't know 100% what exactly happens, I just know how it can be made to not reproduce anymore. Perhaps there was more buggy behaviour causing the actual failure, it would be better to fix that as well.

### CODE REVIEW

**Comment by est31** [comment]
That would be #4138.

**Comment by paramat** [comment]
@nerzhul jenkins isn't building for many recent PRs.

File: src/network/connection.h [hunk 1]
```diff
@@ -687,21 +686,16 @@ class Peer {
 		virtual void PutReliableSendCommand(ConnectionCommand &c,
 						unsigned int max_packet_size) {};
 
-		virtual bool isActive() { return false; };
-
 		virtual bool getAddress(MTProtocols type, Address& toset) = 0;
 
+		bool isPendingDeletion()
+		{ MutexAutoLock lock(m_exclusive_access_mutex); return m_pending_deletion; };
```
Comment (nerzhul) [inline src/network/connection.h]: remove this lock and use atomic please

**Comment by nerzhul** [comment]
Except the commit i did, okay for me.
@paramat it's building but there is a notification problem since recent github api changes as it seems...

**Comment by paramat** [comment]
Ok thanks.

File: src/network/connection.h (hunk 1 above)
Comment (est31) [inline src/network/connection.h]: Can that be done in a separate commit? Its off topic to the PR.

Comment (nerzhul) [inline src/network/connection.h]: @est31 yes, could be done in a separate PR and could be nice to have it faster after this :)

**Comment by nerzhul** [comment]
Just a thing, did you tested it over the WAN to see if the packet re-ordering is not sensible to latencies caused by WAN ?
